consensus/hotstuff/voteaggregator/vote_aggregator.go

```diff
@@ -2,390 +2,170 @@ package voteaggregator
-//   * Provided the vote is valid by itself, treat it as valid. The validity of a vote
-//     needs to be objective and not depend on whether the replica has knowledge of a
-//     conflicting vote.
-//
-// Note that treating a valid vote as invalid would create the following additional edge case:
-//   * consider a primary that is proposing two conflicting blocks (block equivocation)
-//   * Assume the case where we would treat the proposer's vote for its second block
-//     (embedded in the block) as invalid
-//   * then, we would arrive at the conclusion that the second block itself is invalid
-// This would violate objective validity of blocks.
-func (va *VoteAggregator) validateAndStoreIncorporatedVote(vote *model.Vote, block *model.Block) (bool, error) {
-	// validate the vote
-	voter, err := va.voteValidator.ValidateVote(vote, block)
-	if model.IsInvalidVoteError(err) {
-		// does not report invalid vote as an error, notify consumers instead
-		va.notifier.OnInvalidVoteDetected(vote)
-		return false, nil
-	}
-	if err != nil {
-		return false, fmt.Errorf("could not validate incorporated vote: %w", err)
+func (va *VoteAggregator) AddVote(vote *model.Vote) error {
+	// drop stale votes
+	if va.isVoteStale(vote) {
+		return nil
 	}
 
-	// check for double vote:
-	firstVote, detected := va.detectDoubleVote(vote)
-	if detected {
-		va.notifier.OnDoubleVotingDetected(firstVote, vote)
+	// It's ok to silently drop votes in case our processing pipeline is full.
+	// It means that we are probably catching up.
+	if ok := va.pendingVotes.Push(vote); ok {
+		va.pendingVotesNotifier.Notify()
 	}
 
-	// update existing voting status or create a new one
-	votingStatus, exists := va.blockIDToVotingStatus[vote.BlockID]
-	if !exists {
-		// get all identities
-		identities, err := va.committee.Identities(vote.BlockID, filter.Any)
-		if err != nil {
-			return false, fmt.Errorf("error retrieving consensus participants: %w", err)
-		}
+	return nil
+}
 
-		// create VotingStatus for block
-		stakeThreshold := hotstuff.ComputeStakeThresholdForBuildingQC(identities.TotalStake()) // stake threshold for building valid qc
-		votingStatus = NewVotingStatus(block, stakeThreshold, va.signer)
-		va.blockIDToVotingStatus[vote.BlockID] = votingStatus
+func (va *VoteAggregator) AddBlock(block *model.Block) error {
+	// check if the block is for a view that has already been pruned (and is thus stale)
+	if va.isBlockStale(block) {
+		return nil
 	}
-	votingStatus.AddVote(vote, voter)
-	va.updateState(vote)
-	return true, nil
-}
 
-func (va *VoteAggregator) updateState(vote *model.Vote) {
-	voterID := vote.SignerID
+	// TODO: check block signature
 
-	// update viewToVoteID
-	idToVote, exists := va.viewToVoteID[vote.View]
-	if exists {
-		idToVote[voterID] = vote
-	} else {
-		idToVote = make(map[flow.Identifier]*model.Vote)
-		idToVote[voterID] = vote
-		va.viewToVoteID[vote.View] = idToVote
+	err := va.collectors.ProcessBlock(block)
```

Comment:
It's better to get the vote collector then use it to process block.

The processing should verify the block's signature as well as QC, and return the validation result back to the caller (EventHandler), so that the blocked EventHandler can continue the logic of processing the block.

Reply:
`VoteCollector` itself doesn't process blocks `ProcessBlock`s main purpose is to change state of specific collector, we could also change it here as another option.